helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java

```diff
@@ -206,113 +205,29 @@ public void addInstance(String clusterName, InstanceConfig instanceConfig) {
-                  + " has capacity: " + foundInstanceCapacityMap + " and this instance: " + nodeId
-                  + " has capacity: " + instanceCapacityMap);
-        }
-      } else if (foundInstanceConfigsWithMatchingLogicalId.get(0).getInstanceOperation()
-          .equals(InstanceConstants.InstanceOperation.EVACUATE.name())) {
-        // No need to check anything on the new node, the old node will be evacuated and the new node
-        // will be added.
-      } else {
-        // If the instanceConfig.getInstanceEnabled() is true and the existing instance with the same logicalId
-        // does not have InstanceOperation set to one of the above, we cannot add this instance.
-        throw new HelixException(
-            "Instance can only be added if the exising instance sharing the same logicalId"
-                + " has InstanceOperation set to "
-                + InstanceConstants.InstanceOperation.SWAP_OUT.name()
-                + " and this instance has InstanceOperation set to "
-                + InstanceConstants.InstanceOperation.SWAP_IN.name()
-                + " or the existing instance sharing the same logicalId has Instance Operation set to "
-                + InstanceConstants.InstanceOperation.EVACUATE.name()
-                + " and this instance has InstanceOperation unset. Existing instance: "
-                + foundInstanceConfigsWithMatchingLogicalId.get(0).getInstanceName()
-                + " has InstanceOperation: " + foundInstanceConfigsWithMatchingLogicalId.get(0)
-                .getInstanceOperation());
-      }
-    } else if (!instanceConfig.getInstanceOperation().isEmpty()) {
-      // If there are no instances with the same logicalId, we can only add this instance if InstanceOperation
-      // is unset because it is a new instance.
-      throw new HelixException(
-          "There is no instance with logicalId: " + toAddInstanceLogicalId + " in cluster: "
-              + clusterName + "; therefore, " + nodeId
-              + " cannot join cluster with InstanceOperation set to "
-              + instanceConfig.getInstanceOperation() + ".");
+    InstanceConstants.InstanceOperation attemptedInstanceOperation =
+        instanceConfig.getInstanceOperation();
+    try {
+      validateInstanceOperationTransition(instanceConfig,
+          !matchingLogicalIdInstances.isEmpty() ? matchingLogicalIdInstances.get(0) : null,
+          InstanceConstants.InstanceOperation.UNKNOWN,
+          attemptedInstanceOperation, clusterName);
+    } catch (HelixException e) {
+      instanceConfig.setInstanceOperation(InstanceConstants.InstanceOperation.UNKNOWN);
+      logger.error("Failed to add instance " + instanceConfig.getInstanceName() + " to cluster "
+          + clusterName + " with instance operation " + attemptedInstanceOperation
+          + ". Setting INSTANCE_OPERATION to " + instanceConfig.getInstanceOperation()
+          + " instead.", e);
```

Comment:
Be careful with logging, if pipeline consistent happening, then the log with stack trace will blast the server.

Reply:
This will only be called by the participant or externally one time so the logs should not keep being printed.